### PR TITLE
Allow conflicting Git URLs that refer to the same commit SHA

### PR DESCRIPTION
**Pull request by charliermarsh**
## Summary

This PR leverages our lookahead direct URL resolution to significantly improve the range of Git URLs that we can accept (e.g., if a user provides the same requirement, once as a direct dependency, and once as a tag). We did some of this in #2285, but the solution here is more general and works for arbitrary transitive URLs.

Closes https://github.com/astral-sh/uv/issues/2614.

### CODE REVIEW

**Comment by zanieb** [comment]
Nice. I'll give this a real read in the morning.

**Comment by charliermarsh** [comment]
Thanks :)

File: crates/uv/tests/pip_compile.rs [hunk 1]
```diff
@@ -1660,22 +1671,26 @@ fn conflicting_repeated_url_dependency() -> Result<()> {
     ----- stdout -----
 
     ----- stderr -----
-    error: Requirements contain conflicting URLs for package `werkzeug`:
-    - git+https://github.com/pallets/werkzeug.git@2.0.0
-    - git+https://github.com/pallets/werkzeug@3.0.0
+    error: Requirements contain conflicting URLs for package `anyio`:
+    - git+https://github.com/agronholm/anyio.git@4.3.0
+    - git+https://github.com/agronholm/anyio.git@4.0.0
```
Comment (zanieb) [inline crates/uv/tests/pip_compile.rs:1676]: Is there a test case like this for a commit that does not match a tag? Do we display the commit that the tag resolves to in that case?

**Review by zanieb — APPROVED** [review]
Way better! Minor comment regarding a test

File: crates/uv/tests/pip_compile.rs (hunk 1 above)
Comment (charliermarsh) [reply crates/uv/tests/pip_compile.rs:1676]: I don't _think_ so. I could add that, though.

Comment (zanieb) [reply crates/uv/tests/pip_compile.rs:1676]: I'd at least like a snapshot, we could consider improving the output in the future

Comment (charliermarsh) [reply crates/uv/tests/pip_compile.rs:1676]: Sorry, yes, we do already have this (`incompatible_narrowed_url_dependency`).

**Pull request merged**